contracts/upgradeable_contracts/arbitrary_message/BasicForeignAMB.sol

```diff
@@ -0,0 +1,27 @@
+pragma solidity 0.4.24;
+
+import "../../libraries/Message.sol";
+import "../../libraries/ArbitraryMessage.sol";
+import "./BasicAMB.sol";
+import "./MessageDelivery.sol";
+import "./MessageProcessor.sol";
+import "../MessageRelay.sol";
+
+contract BasicForeignAMB is BasicAMB, MessageRelay, MessageDelivery, MessageProcessor {
+    function executeSignatures(bytes _data, uint8[] vs, bytes32[] rs, bytes32[] ss) external {
```

Comment:
I would like to get you opinion here. If we consider as arrays are encoded to be passed as method parameters you will see the following (I took the previous implementation of `executeSignatures` where the message was at the end of the parameters list):
* The first 32 bytes point to the place where `uint8[] vs` is stored.
* The second 32 bytes point to the place where `bytes32[] rs` is stored.
* The third 32 bytes point to the place where `bytes32[] ss` is stored.
* The fourth 32 bytes point to the place where the message begins.
* `uint8[] vs` takes 4*32 bytes, where the first 32 bytes store 3.
* `bytes32[] rs` takes 4*32 bytes, where the first 32 bytes store 3.
* `bytes32[] ss` takes 4*32 bytes, where the first 32 bytes store 3.

This is an [example](https://blockscout.com/eth/mainnet/tx/0x5b0b29bbe1604d5e9bad3bf9af2a615eaffdd946b2569dbad50a4c6937936b1b):
```
0x232a2c1d
000 - 031: 0000000000000000000000000000000000000000000000000000000000000080 - 128
032 - 063: 0000000000000000000000000000000000000000000000000000000000000100 - 256
064 - 095: 0000000000000000000000000000000000000000000000000000000000000180 - 384
096 - 127: 0000000000000000000000000000000000000000000000000000000000000200 - 512
128 - 159: 0000000000000000000000000000000000000000000000000000000000000003 - 3
160 - 191: 000000000000000000000000000000000000000000000000000000000000001c
192 - 223: 000000000000000000000000000000000000000000000000000000000000001b
224 - 255: 000000000000000000000000000000000000000000000000000000000000001c
256 - 287: 0000000000000000000000000000000000000000000000000000000000000003 - 3
288 - 319: e7dfc36bdc721319d943a2d70fcedfac456ca9395ec50404c24178fe08b48837
320 - 351: b85ef14a1a117d9a0781eb07dd0e53e445f14d24629682b7413f23365dcb4f8e
352 - 383: 13bbac2ed6b40a0be695f25b125f610ef6305bfea1cb7651b2d2cb89c45920d9
384 - 415: 0000000000000000000000000000000000000000000000000000000000000003 - 3
416 - 447: 0fba05cf0ba056230150d84975c46bfa961e3181e12efdaa4614de332f6ecf45
448 - 479: 12c95da4c7034a92ad6ee577925e9d9ebf48650c992d5f016193324d4766b56a
480 - 511: 017a79312986afea6350b97a48aec7a6b9956ec8eef2b6e9c48a05da709ef59f
512 - 543: 0000000000000000000000000000000000000000000000000000000000000068 - 104
b96432f3307a72266ab210a8536726d666aa54d0
0000000000000000000000000000000000000000000000001aec43b8b0420000
6eef8dc1865bd54ece19ea9be39996db7e5a44c592f74be7f28d573d2dbf8d99
4aa42145aa6ebf72e164c9bbc74fbd3788045016000000000000000000000000000000000000000000000000
```

Do you think if we pack the arrays before the sending will it consume less gas? I am interesting how much gas it could take to unpack them.

For example, we can use the following method to pack the data: `length + vs1 + vs2 + vs3 + rs1 + rs2 + rs3 + ss1 + ss2 + ss3`. If we pack length and `vs` components as `uint8` it will reduce the transaction input by `3*32+60=156` bytes in case of 3 signatures. It consumes 816 gas.

Reply:
This looks interesting, we can try to run some tests to know if it worth packing the data and unpacking it in the contracts based on the gas usage. This could be used for executing signatures for amb bridge but also for other bridge modes, does it makes sense to create a separate issue for this?

Reply:
I think we can implement this approach in the AMB first. So far, there is no necessity to optimize the original token bridge contracts.

Reply:
Following the idea of packing data as `length + vs1 + vs2 + vs3 + rs1 + rs2 + rs3 + ss1 + ss2 + ss3`
I added an implementation to unpack signature data which I combined with `hasEnoughValidSignatures` method to avoid executing two `for` loops. Probably this can be improved to save more gas.

```solidity
function hasEnoughValidSignaturesAMB(
    bytes _message,
    bytes _signatures,
    IBridgeValidators _validatorContract,
    bool isAMBMessage
) internal view {
    require(isAMBMessage || (!isAMBMessage && isMessageValid(_message)));
    uint256 requiredSignatures = _validatorContract.requiredSignatures();
    uint8 amount;
    assembly {
        amount := mload(add(_signatures, 1))
    }
    require(amount >= requiredSignatures);
    bytes32 hash = hashMessage(_message, isAMBMessage);
    address[] memory encounteredAddresses = new address[](requiredSignatures);

    for (uint256 i = 0; i < requiredSignatures; i++) {
        uint8 v;
        bytes32 r;
        bytes32 s;
        uint256 posr = 33 + amount + 32 * i;
        uint256 poss = posr + 32 * amount;
        assembly {
            v := mload(add(_signatures, add(2, i)))
            r := mload(add(_signatures, posr))
            s := mload(add(_signatures, poss))
        }

        address recoveredAddress = ecrecover(hash, v, r, s);
        require(_validatorContract.isValidator(recoveredAddress));
        require(!addressArrayContains(encounteredAddresses, recoveredAddress));
        encounteredAddresses[i] = recoveredAddress;
    }
}
```

To compare the gas usage, I added a new method and to simplify the comparison, I removed all the logic for processing the message and only tested checking the signatures:

Method with array signatures:
```solidity
function executeSignatures(bytes _data, uint8[] vs, bytes32[] rs, bytes32[] ss) external {
    Message.hasEnoughValidSignatures(_data, vs, rs, ss, validatorContract(), true);
}
```

Method with bytes signatures:
```solidity
function executeSignaturesAMB(bytes _data, bytes _signatures) external {
    Message.hasEnoughValidSignaturesAMB(_data, _signatures, validatorContract(), true);
}
```

I run tests with signatures from 1, 3 and 5 validators, the following were the gas consumed by the transactions:

|Method | 1 | 3 | 5 |
|-----|-----|-----|-----|
array signatures | 50777 | 77334 | 107244
bytes signatures | 49788 | 75624 | 105321

From the results, it looks like packing the signatures has a little better gas usage. Maybe if we can optimize more the method `hasEnoughValidSignaturesAMB` we can get a better result.

Do you think it is worth implementing this logic? Logic on oracle side will need to be updated to reflect this change for AMB mode

Reply:
Cool! Thanks for research! What your expectations how much it could take to introduce changes on both oracle and contracts is?

Reply:
Well, this shouldn't take big efforts. To perform this tests I had to create some helper methods to pack the signature data so I can reuse this for the oracle (plus adding some unit tests) and on the contracts I need to properly include the new already implemented method and update the units tests

Reply:
Applied the changes on the contracts side https://github.com/poanetwork/poa-bridge-contracts/pull/77/commits/0309b3d3739bd9c21b764c41f17e6c18cc506ec3

Reply:
Thanks!